fitbenchmarking/controllers/base_matlab_controller.py

```diff
@@ -0,0 +1,78 @@
+"""
+Implements base class for the matlab fitting software controllers.
+"""
+
+try:
+    from tempfile import TemporaryDirectory
+except ImportError:
+    from backports.tempfile import TemporaryDirectory
+import os
+try:
+    import dill
+    import_success = True
+except ImportError:
+    import_success = False
+
+from fitbenchmarking.controllers.base_controller import Controller
+from fitbenchmarking.utils.exceptions import MissingSoftwareError
+
+
+class BaseMatlabController(Controller):
+    """
+    Base class for matlab fitting software controllers
+    """
+
+    def __init__(self, cost_func):
+        """
+        Initialise anything that is needed specifically for matlab
+        fitting software
+        """
+        super().__init__(cost_func)
+
+        if not import_success:
+            raise MissingSoftwareError('Requirements are missing for Matlab '
+                                       'fitting, module "dill" is reuqired.')
+
+        self.initial_params_mat = None
+
+    def setup(self):
+        raise NotImplementedError
+
+    def jacobian_information(self):
+        raise NotImplementedError
+
+    def fit(self):
+        raise NotImplementedError
+
+    def cleanup(self):
+        raise NotImplementedError
+
+    def py_to_mat(self, func, eng):
+        """
+        Function that serializes a python function and then
+        loads it into the Matlab engine workspace
+        """
+        temp_dir = TemporaryDirectory()
+        temp_file = os.path.join(temp_dir.name, 'temp.pickle')
+        with open(temp_file, 'wb') as f:
+            dill.dump(func, f)
+        eng.workspace['temp_file'] = temp_file
+        eng.evalc('py_f = py.open(temp_file,"rb")')
+        eng.evalc('func_mat = py.dill.load(py_f)')
+        eng.evalc('py_f.close()')
+
+        return eng.workspace['func_mat']
```

Comment:
Since this is now in a helper function it will have a side effect of polluting the namespace, I think we could do with making these variables a bit less likely to cause a clash.

E.g. a new controller could feasibly write:
```
eng.workspace['py_f'] = self.py_to_mat(self._eval_f, eng)
eng.workspace['py_j'] = self.py_to_mat(self._eval_j, eng)
```